controllers/trainingjob/trainingjob_controller.go

```diff
@@ -18,411 +18,347 @@ package trainingjob
-	// BASE CASES
-	//////////////////////////////////////////////////////////////////////////////////////////////////////////
+	//TODO: Convert it to tinyurl or even better can we expose CW url via API server proxy UI?
+	ctx.CloudWatchLogURL = "https://" + *ctx.TrainingJob.Spec.Region + ".console.aws.amazon.com/cloudwatch/home?region=" +
+		*ctx.TrainingJob.Spec.Region + "#ctx.LogStream:group=/aws/sagemaker/TrainingJobs;prefix=" +
+		*ctx.TrainingJobDescription.TrainingJobName + ";streamFilter=typeLogStreamPrefix"
 
-	switch trainingJobDescription.TrainingJobStatus {
+	switch ctx.TrainingJobDescription.TrainingJobStatus {
 	case sagemaker.TrainingJobStatusInProgress, sagemaker.TrainingJobStatusStopping:
-		if err = r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
+		if !controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
+			if err = r.handleUpdates(ctx); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to update SageMaker trainingjob"))
+			}
+		} else {
+			if _, err := ctx.SageMakerClient.StopTrainingJob(ctx, ctx.TrainingJobName); err != nil {
+				return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Unable to delete training job"))
+			}
 		}
-		return RequeueAfterInterval(r.PollInterval, err)
 
-	case sagemaker.TrainingJobStatusStopped, sagemaker.TrainingJobStatusFailed:
-		return NoRequeue()
+		break
 
 	case sagemaker.TrainingJobStatusCompleted:
-		// If job has completed populate the model full path
-		log.Info("Training has completed updating model path")
-		// SageMaker stores the model artifact in OutputDataConfig path with path /output/model.tar.gz
-		// SageMaker documentation https://docs.aws.amazon.com/sagemaker/latest/dg/cdf-training.html
-		const outputPath string = "/output/model.tar.gz"
-		modelPath := *state.Spec.OutputDataConfig.S3OutputPath + state.Status.SageMakerTrainingJobName + outputPath
-		if err = r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-			SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-			TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-			SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-			LastCheckTime:            Now(),
-			CloudWatchLogUrl:         cwLogUrl,
-			ModelPath:                modelPath,
-		}); err != nil {
-			log.Info("Error updating ETCD to sync with SM API state")
-			return RequeueIfError(err)
+		if err = r.updateCompleted(ctx); err != nil {
```

Comment:
This name doesn't do a good job of describing what the function does. Consider `addModelPathToStatus`?